synapse/rest/client/v1/room.py

```diff
@@ -38,6 +38,7 @@
     parse_integer,
```

Comment:
(moving this discussion to a thread)

@erikjohnston When using the `backfilled` option, something is going wonky with the `/messages?dir=b` ordering. They should be newest to oldest. We are backfilling the messages next to `eventsBefore`.

The actual messages response starts out with `eventsAfter` as expected but then goes straight to `eventsBefore` instead of the `backfilledEvents`.

If I remove the [`backfilled` changes](https://github.com/matrix-org/synapse/pull/9247#issuecomment-782559083), the ordering is correct and as expected.

Any pointers for why it's returning backfilled out of order?

I've marked the `backfilledEvents` with the diff red(`-`) and the `eventsBefore` as green(`+`) just to group them a bit better and make it a little bit more clear. It's not an actual diff, just using the colors.

 - **Expected Events** from `/messages?dir=b` (newest to oldest): `[$cr8O0dQGwdj4HLbKtq_OnnHYbu5MYDRUHbtM8ginimk $b-tslZ2gKFZAG45p89d1x4WVO4DPg7DtBMQZTmVEHo4 $K-1YLxBGXMMCy1KcSD8mSJhDic5Mxfu93SYbTmvLMlc $qzwcw274C8_uL7wM1a5XAd49SHuUpkvbfkTK14T626E $OPHISXPJn99vW5KkUaKguW1Ik1Ev8UCObMpUv3QA-jI $GU1U44caIFzNLz7x_HdVELidHKsbxGHDmu5V2qI3-ns]`
    ```diff
    [
        $cr8O0dQGwdj4HLbKtq_OnnHYbu5MYDRUHbtM8ginimk
        $b-tslZ2gKFZAG45p89d1x4WVO4DPg7DtBMQZTmVEHo4
    -    $K-1YLxBGXMMCy1KcSD8mSJhDic5Mxfu93SYbTmvLMlc
    -    $qzwcw274C8_uL7wM1a5XAd49SHuUpkvbfkTK14T626E
    -    $OPHISXPJn99vW5KkUaKguW1Ik1Ev8UCObMpUv3QA-jI
    +    $GU1U44caIFzNLz7x_HdVELidHKsbxGHDmu5V2qI3-ns
    ]
    ```
 - **Actual Events** from `/messages?dir=b`: `[$cr8O0dQGwdj4HLbKtq_OnnHYbu5MYDRUHbtM8ginimk $b-tslZ2gKFZAG45p89d1x4WVO4DPg7DtBMQZTmVEHo4 $GU1U44caIFzNLz7x_HdVELidHKsbxGHDmu5V2qI3-ns $OPHISXPJn99vW5KkUaKguW1Ik1Ev8UCObMpUv3QA-jI $qzwcw274C8_uL7wM1a5XAd49SHuUpkvbfkTK14T626E $K-1YLxBGXMMCy1KcSD8mSJhDic5Mxfu93SYbTmvLMlc]`
    ```diff
     [
        $cr8O0dQGwdj4HLbKtq_OnnHYbu5MYDRUHbtM8ginimk
        $b-tslZ2gKFZAG45p89d1x4WVO4DPg7DtBMQZTmVEHo4
    +    $GU1U44caIFzNLz7x_HdVELidHKsbxGHDmu5V2qI3-ns
    -    $OPHISXPJn99vW5KkUaKguW1Ik1Ev8UCObMpUv3QA-jI
    -    $qzwcw274C8_uL7wM1a5XAd49SHuUpkvbfkTK14T626E
    -    $K-1YLxBGXMMCy1KcSD8mSJhDic5Mxfu93SYbTmvLMlc
    ]
    ```

Here are the different pieces (ordered from oldest to newest):

 - `eventsBefore` (in the + green): `[$GU1U44caIFzNLz7x_HdVELidHKsbxGHDmu5V2qI3-ns]`
 - `backfilledEvents` (in the - red): `[$OPHISXPJn99vW5KkUaKguW1Ik1Ev8UCObMpUv3QA-jI $qzwcw274C8_uL7wM1a5XAd49SHuUpkvbfkTK14T626E $K-1YLxBGXMMCy1KcSD8mSJhDic5Mxfu93SYbTmvLMlc]`
 - `eventsAfter`: `[$b-tslZ2gKFZAG45p89d1x4WVO4DPg7DtBMQZTmVEHo4 $cr8O0dQGwdj4HLbKtq_OnnHYbu5MYDRUHbtM8ginimk]`

To see what the Complement test is doing, see the `Backfilled messages come back in correct order` test in https://github.com/matrix-org/complement/pull/68/files

---

The DAG looks correct:

![](https://user-images.githubusercontent.com/558581/108966126-6a1b6a00-7643-11eb-860a-04435d5285b2.png)


Here is the `/context` response I used to generate that visualization in TARDIS (plopped this in https://github.com/matrix-org/tardis/pull/1): 
<details>

<summary>GET /_matrix/client/r0/rooms/{roomId}/context/{eventId}</summary>

```json
{
  "events_before": [
    {
      "auth_events": [
        "$U4bQ35C3w8D271raR3K1pBT74kMPPHNwVzIz0AVTODE",
        "$ADZFDE6McQzy-WiZw0daxLb8g9iulClp-Rvm81S96Gs",
        "$iGaP7cnnC0NQ1R3UX-HcyKL3lPxUxMxLWy-ccrtnlo0"
      ],
      "prev_events": [
        "$GU1U44caIFzNLz7x_HdVELidHKsbxGHDmu5V2qI3-ns"
      ],
      "type": "m.room.message",
      "room_id": "!zbFaRylcNixoifxaEF:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message 0",
        "msgtype": "m.text"
      },
      "depth": 8,
      "prev_state": [],
      "origin": "hs1",
      "origin_server_ts": 1614151637180,
      "hashes": {
        "sha256": "4ykbcCLATbyHMxwBR1BYcaNVbMuLWIHLJpt3XWnTKss"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_aHRx": "VUY/H0TPMK6TI5Dd5IOXSz4WCpB7Y7P6VcW8dhjDb+FC9nJjLXyaF8u0u0/dALyWatewgT1yZl8a90HYZLrSBQ"
        }
      },
      "unsigned": {
        "age": 477
      },
      "event_id": "$b-tslZ2gKFZAG45p89d1x4WVO4DPg7DtBMQZTmVEHo4",
      "user_id": "@alice:hs1",
      "age": 477
    },
    {
      "auth_events": [
        "$U4bQ35C3w8D271raR3K1pBT74kMPPHNwVzIz0AVTODE",
        "$ADZFDE6McQzy-WiZw0daxLb8g9iulClp-Rvm81S96Gs",
        "$iGaP7cnnC0NQ1R3UX-HcyKL3lPxUxMxLWy-ccrtnlo0"
      ],
      "prev_events": [
        "$ADZFDE6McQzy-WiZw0daxLb8g9iulClp-Rvm81S96Gs"
      ],
      "type": "m.room.message",
      "room_id": "!zbFaRylcNixoifxaEF:hs1",
      "sender": "@alice:hs1",
      "content": {
        "body": "Message 0",
        "msgtype": "m.text"
      },
      "depth": 7,
      "prev_state": [],
      "origin": "hs1",
      "origin_server_ts": 1614151636919,
      "hashes": {
        "sha256": "L+O2d2VA1Jm8pcgGf7830hhdyLbUxN+UVDwZUyer8Js"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_aHRx": "LEe8H4cEP3uRHO9TuYJuM9euq0xhglib/Pds1zrrZzVIx+aVnlbhIpCD3SqfXs5y8D4j2rJUamWC+DqFiirSAw"
        }
      },
      "unsigned": {
        "age": 738
      },
      "event_id": "$GU1U44caIFzNLz7x_HdVELidHKsbxGHDmu5V2qI3-ns",
      "user_id": "@alice:hs1",
      "age": 738
    },
    {
      "auth_events": [
        "$U4bQ35C3w8D271raR3K1pBT74kMPPHNwVzIz0AVTODE",
        "$GGZD3jkbL9QwnRCqQaBtOcIEfLVjA_62e9O6LseW6ZQ",
        "$iGaP7cnnC0NQ1R3UX-HcyKL3lPxUxMxLWy-ccrtnlo0"
      ],
      "prev_events": [
        "$GU1U44caIFzNLz7x_HdVELidHKsbxGHDmu5V2qI3-ns"
      ],
      "type": "m.room.message",
      "room_id": "!zbFaRylcNixoifxaEF:hs1",
      "sender": "@the-bridge-user:hs1",
      "content": {
        "body": "Backfilled 0",
        "m.historical": true,
        "msgtype": "m.text"
      },
      "depth": 7,
      "prev_state": [],
      "origin_server_ts": 1614151637151,
      "origin": "hs1",
      "hashes": {
        "sha256": "Zd/bUoNGegf6uzAKzpGT4D1PrNYMneqCw7zhvNZDB1Q"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_aHRx": "u4fX9GjavkE/s37ST8LKyi/8oFUVoc7xNddnzJUSascXFWMSmGk/7RWgZJ6v1qPRj7CP9lWh0yKE5hfNdRrgCg"
        }
      },
      "unsigned": {
        "age": 149
      },
      "event_id": "$OPHISXPJn99vW5KkUaKguW1Ik1Ev8UCObMpUv3QA-jI",
      "user_id": "@the-bridge-user:hs1",
      "age": 149
    },
    {
      "auth_events": [
        "$U4bQ35C3w8D271raR3K1pBT74kMPPHNwVzIz0AVTODE",
        "$GGZD3jkbL9QwnRCqQaBtOcIEfLVjA_62e9O6LseW6ZQ",
        "$iGaP7cnnC0NQ1R3UX-HcyKL3lPxUxMxLWy-ccrtnlo0"
      ],
      "prev_events": [
        "$OPHISXPJn99vW5KkUaKguW1Ik1Ev8UCObMpUv3QA-jI"
      ],
      "type": "m.room.message",
      "room_id": "!zbFaRylcNixoifxaEF:hs1",
      "sender": "@the-bridge-user:hs1",
      "content": {
        "body": "Backfilled 1",
        "m.historical": true,
        "msgtype": "m.text"
      },
      "depth": 7,
      "prev_state": [],
      "origin_server_ts": 1614151637152,
      "origin": "hs1",
      "hashes": {
        "sha256": "3Vz1oDN+pS4iiJatP+vhP+N8aH3R49I+iiK7aRb4SWs"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_aHRx": "NQAhIggubxxUQlzDOR8duW1NH6eAIRHp09I0+7RD1BOF3HJDqEnTN/xE1vQ5bf5ypzCYVRaODlNe+KMnexoLDg"
        }
      },
      "unsigned": {
        "age": 102
      },
      "event_id": "$qzwcw274C8_uL7wM1a5XAd49SHuUpkvbfkTK14T626E",
      "user_id": "@the-bridge-user:hs1",
      "age": 102
    },
    {
      "auth_events": [
        "$U4bQ35C3w8D271raR3K1pBT74kMPPHNwVzIz0AVTODE",
        "$GGZD3jkbL9QwnRCqQaBtOcIEfLVjA_62e9O6LseW6ZQ",
        "$iGaP7cnnC0NQ1R3UX-HcyKL3lPxUxMxLWy-ccrtnlo0"
      ],
      "prev_events": [
        "$qzwcw274C8_uL7wM1a5XAd49SHuUpkvbfkTK14T626E"
      ],
      "type": "m.room.message",
      "room_id": "!zbFaRylcNixoifxaEF:hs1",
      "sender": "@the-bridge-user:hs1",
      "content": {
        "body": "Backfilled 2",
        "m.historical": true,
        "msgtype": "m.text"
      },
      "depth": 7,
      "prev_state": [],
      "origin_server_ts": 1614151637153,
      "origin": "hs1",
      "hashes": {
        "sha256": "rB/q7D4my6pTIfRTyPIPPcs/wCMdMsEBwXA/bs4sXXM"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_aHRx": "+zn0sLLROpLMF0zHBm0sXkpV16Oj/HXoRIxnCgZy9tXxByrvmyfOSf/aDja38xQpWmu1Q9+TcQSvD6KrcVY4DQ"
        }
      },
      "unsigned": {
        "age": 69
      },
      "event_id": "$K-1YLxBGXMMCy1KcSD8mSJhDic5Mxfu93SYbTmvLMlc",
      "user_id": "@the-bridge-user:hs1",
      "age": 69
    },
    {
      "auth_events": [
        "$U4bQ35C3w8D271raR3K1pBT74kMPPHNwVzIz0AVTODE",
        "$iGaP7cnnC0NQ1R3UX-HcyKL3lPxUxMxLWy-ccrtnlo0",
        "$1ULXaSRrFQuWUCxbtMYUXJm8mJKZV2Upy7pKynFpcfc"
      ],
      "prev_events": [
        "$-Yi5y34ocqpGZIkqg1vVbMVmJXUFSJZZyqj1tK1kyE8"
      ],
      "type": "m.room.member",
      "room_id": "!zbFaRylcNixoifxaEF:hs1",
      "sender": "@alice:hs1",
      "content": {
        "membership": "join",
        "displayname": "alice"
      },
      "depth": 6,
      "prev_state": [],
      "state_key": "@alice:hs1",
      "origin": "hs1",
      "origin_server_ts": 1614151636770,
      "hashes": {
        "sha256": "3/fkoBWOKAsMVSyGpcZMSVFAOCZMXt/H5s1YYDJga8c"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_aHRx": "jp0LOSWrANOEF6MMlstGxH0UFbwe+DI76gd0nA4/aAOk+FaLqqZZvHROjngSUobHcT8LYy4bq7aeNt+GngZEBQ"
        }
      },
      "unsigned": {
        "age": 887
      },
      "event_id": "$ADZFDE6McQzy-WiZw0daxLb8g9iulClp-Rvm81S96Gs",
      "user_id": "@alice:hs1",
      "age": 887
    },
    {
      "auth_events": [
        "$U4bQ35C3w8D271raR3K1pBT74kMPPHNwVzIz0AVTODE",
        "$GGZD3jkbL9QwnRCqQaBtOcIEfLVjA_62e9O6LseW6ZQ",
        "$iGaP7cnnC0NQ1R3UX-HcyKL3lPxUxMxLWy-ccrtnlo0"
      ],
      "prev_events": [
        "$1ULXaSRrFQuWUCxbtMYUXJm8mJKZV2Upy7pKynFpcfc"
      ],
      "type": "m.room.history_visibility",
      "room_id": "!zbFaRylcNixoifxaEF:hs1",
      "sender": "@the-bridge-user:hs1",
      "content": {
        "history_visibility": "shared"
      },
      "depth": 5,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1614151636654,
      "hashes": {
        "sha256": "ndUCUkpNwRSHN49gkrnjedGuYyHC7icq1g5nl6tdQbU"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_aHRx": "lfNrCzKq9AxhwmCZq6h0jauZcd/fb0vyuacCZwQQ3A9cafrm17g1PenRUchyi448R/b1a72uXe6qJCh+jmHyDQ"
        }
      },
      "unsigned": {
        "age": 1003
      },
      "event_id": "$-Yi5y34ocqpGZIkqg1vVbMVmJXUFSJZZyqj1tK1kyE8",
      "user_id": "@the-bridge-user:hs1",
      "age": 1003
    },
    {
      "auth_events": [
        "$U4bQ35C3w8D271raR3K1pBT74kMPPHNwVzIz0AVTODE",
        "$GGZD3jkbL9QwnRCqQaBtOcIEfLVjA_62e9O6LseW6ZQ",
        "$iGaP7cnnC0NQ1R3UX-HcyKL3lPxUxMxLWy-ccrtnlo0"
      ],
      "prev_events": [
        "$iGaP7cnnC0NQ1R3UX-HcyKL3lPxUxMxLWy-ccrtnlo0"
      ],
      "type": "m.room.join_rules",
      "room_id": "!zbFaRylcNixoifxaEF:hs1",
      "sender": "@the-bridge-user:hs1",
      "content": {
        "join_rule": "public"
      },
      "depth": 4,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1614151636586,
      "hashes": {
        "sha256": "FOamTr4YdxoyEnN4cGSjxsXltTWz4HAh4E6ks8EWPiU"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_aHRx": "nVj+XGhqjf5w5GB93AHwMjYbBLyHSx3regcy1vh7NA08zqKiJBgbW8OK+t/6OcARr82S14p6OvFvDJIC7tZsAA"
        }
      },
      "unsigned": {
        "age": 1071
      },
      "event_id": "$1ULXaSRrFQuWUCxbtMYUXJm8mJKZV2Upy7pKynFpcfc",
      "user_id": "@the-bridge-user:hs1",
      "age": 1071
    },
    {
      "auth_events": [
        "$U4bQ35C3w8D271raR3K1pBT74kMPPHNwVzIz0AVTODE",
        "$GGZD3jkbL9QwnRCqQaBtOcIEfLVjA_62e9O6LseW6ZQ"
      ],
      "prev_events": [
        "$GGZD3jkbL9QwnRCqQaBtOcIEfLVjA_62e9O6LseW6ZQ"
      ],
      "type": "m.room.power_levels",
      "room_id": "!zbFaRylcNixoifxaEF:hs1",
      "sender": "@the-bridge-user:hs1",
      "content": {
        "users": {
          "@the-bridge-user:hs1": 100
        },
        "users_default": 0,
        "events": {
          "m.room.name": 50,
          "m.room.power_levels": 100,
          "m.room.history_visibility": 100,
          "m.room.canonical_alias": 50,
          "m.room.avatar": 50,
          "m.room.tombstone": 100,
          "m.room.server_acl": 100,
          "m.room.encryption": 100
        },
        "events_default": 0,
        "state_default": 50,
        "ban": 50,
        "kick": 50,
        "redact": 50,
        "invite": 50
      },
      "depth": 3,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1614151636515,
      "hashes": {
        "sha256": "WOCmHmFgHQMfKEcgdCF9Y6f+ANpRWKFeU7NfFTF8lPE"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_aHRx": "nDwuBfFuzllk+Y1HsCwW5gxld7mMEQOcGgIjNCXgDYy1UiRm1h5OP+88sXFuoe6G/UjxKunGj8veQN9YpeJjAA"
        }
      },
      "unsigned": {
        "age": 1142
      },
      "event_id": "$iGaP7cnnC0NQ1R3UX-HcyKL3lPxUxMxLWy-ccrtnlo0",
      "user_id": "@the-bridge-user:hs1",
      "age": 1142
    },
    {
      "auth_events": [
        "$U4bQ35C3w8D271raR3K1pBT74kMPPHNwVzIz0AVTODE"
      ],
      "prev_events": [
        "$U4bQ35C3w8D271raR3K1pBT74kMPPHNwVzIz0AVTODE"
      ],
      "type": "m.room.member",
      "room_id": "!zbFaRylcNixoifxaEF:hs1",
      "sender": "@the-bridge-user:hs1",
      "content": {
        "membership": "join"
      },
      "depth": 2,
      "prev_state": [],
      "state_key": "@the-bridge-user:hs1",
      "origin": "hs1",
      "origin_server_ts": 1614151636429,
      "hashes": {
        "sha256": "40epY96XO8RU3Y3wE9zF+1cA/fmv/v6d0ozUFLYgw+4"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_aHRx": "O+mRbtFNHqR3v6e2bF7ctoHMCW3uTfUgf0XIbexr9F7k+YHuuG20+8LK4h3Dzes4B3D+8ES92ayKJF5dRDHnBg"
        }
      },
      "unsigned": {
        "age": 1228
      },
      "event_id": "$GGZD3jkbL9QwnRCqQaBtOcIEfLVjA_62e9O6LseW6ZQ",
      "user_id": "@the-bridge-user:hs1",
      "age": 1228
    },
    {
      "auth_events": [],
      "prev_events": [],
      "type": "m.room.create",
      "room_id": "!zbFaRylcNixoifxaEF:hs1",
      "sender": "@the-bridge-user:hs1",
      "content": {
        "room_version": "6",
        "creator": "@the-bridge-user:hs1"
      },
      "depth": 1,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1614151636319,
      "hashes": {
        "sha256": "asaL3enzEF4yfUgev5xknSZxfU8zqUlHO///wM89wRc"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_aHRx": "6UU9eUUl+Xmm4/4BTD8h39jxQpNcm9njpHjpT+VZQPEI3s9Pwnym9V8fT9ObX9/pQhl+zX+ogrfUjxm412HTBQ"
        }
      },
      "unsigned": {
        "age": 1338
      },
      "event_id": "$U4bQ35C3w8D271raR3K1pBT74kMPPHNwVzIz0AVTODE",
      "user_id": "@the-bridge-user:hs1",
      "age": 1338
    }
  ],
  "events_after": [],
  "start": "t1-3_0_0_0_0_0_0_0_0",
  "end": "t9-26_0_0_0_0_0_0_0_0",
  "event": {
    "auth_events": [
      "$U4bQ35C3w8D271raR3K1pBT74kMPPHNwVzIz0AVTODE",
      "$ADZFDE6McQzy-WiZw0daxLb8g9iulClp-Rvm81S96Gs",
      "$iGaP7cnnC0NQ1R3UX-HcyKL3lPxUxMxLWy-ccrtnlo0"
    ],
    "prev_events": [
      "$b-tslZ2gKFZAG45p89d1x4WVO4DPg7DtBMQZTmVEHo4"
    ],
    "type": "m.room.message",
    "room_id": "!zbFaRylcNixoifxaEF:hs1",
    "sender": "@alice:hs1",
    "content": {
      "body": "Message 1",
      "msgtype": "m.text"
    },
    "depth": 9,
    "prev_state": [],
    "origin": "hs1",
    "origin_server_ts": 1614151637371,
    "hashes": {
      "sha256": "U0tsJZfK8e2J6M8JoU/QguH7ZsApdhjN3PmjICdBKWA"
    },
    "signatures": {
      "hs1": {
        "ed25519:a_aHRx": "FLiYfu3/qnRkVPGqtUwAd/CUV0X6mBVEPd0FLQVNSC3lEIPp8K5HbxvVhCqFDPLTngASVvvMPjsCCxpT7eUEBw"
      }
    },
    "unsigned": {
      "age": 286
    },
    "event_id": "$cr8O0dQGwdj4HLbKtq_OnnHYbu5MYDRUHbtM8ginimk",
    "user_id": "@alice:hs1",
    "age": 286
  },
  "state": [
    {
      "auth_events": [],
      "prev_events": [],
      "type": "m.room.create",
      "room_id": "!zbFaRylcNixoifxaEF:hs1",
      "sender": "@the-bridge-user:hs1",
      "content": {
        "room_version": "6",
        "creator": "@the-bridge-user:hs1"
      },
      "depth": 1,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1614151636319,
      "hashes": {
        "sha256": "asaL3enzEF4yfUgev5xknSZxfU8zqUlHO///wM89wRc"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_aHRx": "6UU9eUUl+Xmm4/4BTD8h39jxQpNcm9njpHjpT+VZQPEI3s9Pwnym9V8fT9ObX9/pQhl+zX+ogrfUjxm412HTBQ"
        }
      },
      "unsigned": {
        "age": 1338
      },
      "event_id": "$U4bQ35C3w8D271raR3K1pBT74kMPPHNwVzIz0AVTODE",
      "user_id": "@the-bridge-user:hs1",
      "age": 1338
    },
    {
      "auth_events": [
        "$U4bQ35C3w8D271raR3K1pBT74kMPPHNwVzIz0AVTODE",
        "$GGZD3jkbL9QwnRCqQaBtOcIEfLVjA_62e9O6LseW6ZQ"
      ],
      "prev_events": [
        "$GGZD3jkbL9QwnRCqQaBtOcIEfLVjA_62e9O6LseW6ZQ"
      ],
      "type": "m.room.power_levels",
      "room_id": "!zbFaRylcNixoifxaEF:hs1",
      "sender": "@the-bridge-user:hs1",
      "content": {
        "users": {
          "@the-bridge-user:hs1": 100
        },
        "users_default": 0,
        "events": {
          "m.room.name": 50,
          "m.room.power_levels": 100,
          "m.room.history_visibility": 100,
          "m.room.canonical_alias": 50,
          "m.room.avatar": 50,
          "m.room.tombstone": 100,
          "m.room.server_acl": 100,
          "m.room.encryption": 100
        },
        "events_default": 0,
        "state_default": 50,
        "ban": 50,
        "kick": 50,
        "redact": 50,
        "invite": 50
      },
      "depth": 3,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1614151636515,
      "hashes": {
        "sha256": "WOCmHmFgHQMfKEcgdCF9Y6f+ANpRWKFeU7NfFTF8lPE"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_aHRx": "nDwuBfFuzllk+Y1HsCwW5gxld7mMEQOcGgIjNCXgDYy1UiRm1h5OP+88sXFuoe6G/UjxKunGj8veQN9YpeJjAA"
        }
      },
      "unsigned": {
        "age": 1142
      },
      "event_id": "$iGaP7cnnC0NQ1R3UX-HcyKL3lPxUxMxLWy-ccrtnlo0",
      "user_id": "@the-bridge-user:hs1",
      "age": 1142
    },
    {
      "auth_events": [
        "$U4bQ35C3w8D271raR3K1pBT74kMPPHNwVzIz0AVTODE",
        "$GGZD3jkbL9QwnRCqQaBtOcIEfLVjA_62e9O6LseW6ZQ",
        "$iGaP7cnnC0NQ1R3UX-HcyKL3lPxUxMxLWy-ccrtnlo0"
      ],
      "prev_events": [
        "$iGaP7cnnC0NQ1R3UX-HcyKL3lPxUxMxLWy-ccrtnlo0"
      ],
      "type": "m.room.join_rules",
      "room_id": "!zbFaRylcNixoifxaEF:hs1",
      "sender": "@the-bridge-user:hs1",
      "content": {
        "join_rule": "public"
      },
      "depth": 4,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1614151636586,
      "hashes": {
        "sha256": "FOamTr4YdxoyEnN4cGSjxsXltTWz4HAh4E6ks8EWPiU"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_aHRx": "nVj+XGhqjf5w5GB93AHwMjYbBLyHSx3regcy1vh7NA08zqKiJBgbW8OK+t/6OcARr82S14p6OvFvDJIC7tZsAA"
        }
      },
      "unsigned": {
        "age": 1071
      },
      "event_id": "$1ULXaSRrFQuWUCxbtMYUXJm8mJKZV2Upy7pKynFpcfc",
      "user_id": "@the-bridge-user:hs1",
      "age": 1071
    },
    {
      "auth_events": [
        "$U4bQ35C3w8D271raR3K1pBT74kMPPHNwVzIz0AVTODE",
        "$GGZD3jkbL9QwnRCqQaBtOcIEfLVjA_62e9O6LseW6ZQ",
        "$iGaP7cnnC0NQ1R3UX-HcyKL3lPxUxMxLWy-ccrtnlo0"
      ],
      "prev_events": [
        "$1ULXaSRrFQuWUCxbtMYUXJm8mJKZV2Upy7pKynFpcfc"
      ],
      "type": "m.room.history_visibility",
      "room_id": "!zbFaRylcNixoifxaEF:hs1",
      "sender": "@the-bridge-user:hs1",
      "content": {
        "history_visibility": "shared"
      },
      "depth": 5,
      "prev_state": [],
      "state_key": "",
      "origin": "hs1",
      "origin_server_ts": 1614151636654,
      "hashes": {
        "sha256": "ndUCUkpNwRSHN49gkrnjedGuYyHC7icq1g5nl6tdQbU"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_aHRx": "lfNrCzKq9AxhwmCZq6h0jauZcd/fb0vyuacCZwQQ3A9cafrm17g1PenRUchyi448R/b1a72uXe6qJCh+jmHyDQ"
        }
      },
      "unsigned": {
        "age": 1003
      },
      "event_id": "$-Yi5y34ocqpGZIkqg1vVbMVmJXUFSJZZyqj1tK1kyE8",
      "user_id": "@the-bridge-user:hs1",
      "age": 1003
    },
    {
      "auth_events": [
        "$U4bQ35C3w8D271raR3K1pBT74kMPPHNwVzIz0AVTODE"
      ],
      "prev_events": [
        "$U4bQ35C3w8D271raR3K1pBT74kMPPHNwVzIz0AVTODE"
      ],
      "type": "m.room.member",
      "room_id": "!zbFaRylcNixoifxaEF:hs1",
      "sender": "@the-bridge-user:hs1",
      "content": {
        "membership": "join"
      },
      "depth": 2,
      "prev_state": [],
      "state_key": "@the-bridge-user:hs1",
      "origin": "hs1",
      "origin_server_ts": 1614151636429,
      "hashes": {
        "sha256": "40epY96XO8RU3Y3wE9zF+1cA/fmv/v6d0ozUFLYgw+4"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_aHRx": "O+mRbtFNHqR3v6e2bF7ctoHMCW3uTfUgf0XIbexr9F7k+YHuuG20+8LK4h3Dzes4B3D+8ES92ayKJF5dRDHnBg"
        }
      },
      "unsigned": {
        "age": 1228
      },
      "event_id": "$GGZD3jkbL9QwnRCqQaBtOcIEfLVjA_62e9O6LseW6ZQ",
      "user_id": "@the-bridge-user:hs1",
      "age": 1228
    },
    {
      "auth_events": [
        "$U4bQ35C3w8D271raR3K1pBT74kMPPHNwVzIz0AVTODE",
        "$iGaP7cnnC0NQ1R3UX-HcyKL3lPxUxMxLWy-ccrtnlo0",
        "$1ULXaSRrFQuWUCxbtMYUXJm8mJKZV2Upy7pKynFpcfc"
      ],
      "prev_events": [
        "$-Yi5y34ocqpGZIkqg1vVbMVmJXUFSJZZyqj1tK1kyE8"
      ],
      "type": "m.room.member",
      "room_id": "!zbFaRylcNixoifxaEF:hs1",
      "sender": "@alice:hs1",
      "content": {
        "membership": "join",
        "displayname": "alice"
      },
      "depth": 6,
      "prev_state": [],
      "state_key": "@alice:hs1",
      "origin": "hs1",
      "origin_server_ts": 1614151636770,
      "hashes": {
        "sha256": "3/fkoBWOKAsMVSyGpcZMSVFAOCZMXt/H5s1YYDJga8c"
      },
      "signatures": {
        "hs1": {
          "ed25519:a_aHRx": "jp0LOSWrANOEF6MMlstGxH0UFbwe+DI76gd0nA4/aAOk+FaLqqZZvHROjngSUobHcT8LYy4bq7aeNt+GngZEBQ"
        }
      },
      "unsigned": {
        "age": 887
      },
      "event_id": "$ADZFDE6McQzy-WiZw0daxLb8g9iulClp-Rvm81S96Gs",
      "user_id": "@alice:hs1",
      "age": 887
    }
  ]
}
```

</details>

Reply:
@erikjohnston Friendly ping to have a glance at this for any insight. Or if you want me to dive deeper down the `backfilled` path to investigate, etc.

Reply:
Sorry, I'm doing a terrible job at keeping on top of this. Feel free to review request me as well, as I do try and keep on top of my review queue, unlike the tire fire that is my GH notifications....

So I think this is because we order backfilled events *before* normal events at the same depth. In Synapse each event gets assigned a tuple, `(topological_ordering, stream_ordering)`, where `topological_ordering` is just the `depth` and `stream_ordering` is an auto incrementing integer, but for backfilled events its negative. Since the stream ordering is basically insertion ordering, that is the order we use for sending stuff down sync (and since backfilled events have a negative stream ordering they *don't* go down sync). For pagination we order by the tuple, i.e. by depth with stream ordering as the tie break.

So what you're seeing is that the backfilled have high, but negative stream ordering, since they're a) backfilled and b) inserted after `eventsBefore`, and so since the backfilled events have the same depth as `eventsBefore` it gets inserted *before* `eventsBefore`. 

So I think the correct thing is for them to have the same depth as the event they should go before, and a greater depth than those it should come after. I think?

Reply:
Thanks for the fantastic explanation @erikjohnston! 😺 Just got back from a Colorado trip and I've got this implemented now 🚀

Since each backfilled event `stream_ordering` is decremented, the order of the historical messages is backwards when the current Complement tests tries to insert from the oldest history to newest. But this just means it is the opposite now and **there is a reverse-chronological constraint when importing historical messages**. **Just need to refactor the Complement tests to insert from most recent history to oldest.**

---

For reference, in my original [break-down description](https://github.com/matrix-org/matrix-doc/pull/2716#discussion_r559846587) it was the opposite, "there will be a chronological constraint when importing history. So once you import the first chunk of history, you can't import another piece earlier.". But either way works :+1: so reverse-chronological will work just fine!

And around removing this constraint, from the same break-down, "Once we refactor the Synapse/Dendrite codebases to calculate depth ourselves(instead of relying on assuming it from the stream ID), we can remove this constraint. This is a huge undertaking and probably won't be done for this MSC directly."

Reply:
Actually diving into the Complement tests now, the reverse-chronological constraint seems like a problem. From the most recent historical message, I can't reference the previous message because it hasn't been created yet 🤔 

@erikjohnston Is there a way around this?

---

To illustrate the problem if we insert chronological while using the `backfilled` Synapse implementation. The backfilled events are in the opposite order:

```
ActualEvents: [
	$c6TJxAw9Gp9r3HCN7xwK13bOYM9sKxD2q0A8XQ9qZN0 (Message 1 - eventsAfter)
	$G65mlrTJcprHXOXyonpcP-2fBHmvXe133dlW_fmKZes (Message 0 - eventsAfter)
	$njJCUsvrVaPtpyNP6PxW-PFuj9YQdowH3PVgC73wi44 (Backfilled 0)
	$v2Iq9Ih4OyKReqJ2pcTDRb30HrrHEmG0qnhta47Crus (Backfilled 1)
	$cnVNVEDS7cSSN-m0rgfNAZOhfs8yey72rKbJf1zfe5U (Backfilled 2)
	$hSNUYZq2RQVdjgRpXpq4EF4KnI1mU1eDdq5KncNkHak (Message 0 - eventsBefore)
]

ExpectedEvents: [
	$c6TJxAw9Gp9r3HCN7xwK13bOYM9sKxD2q0A8XQ9qZN0 (Message 1 - eventsAfter)
	$G65mlrTJcprHXOXyonpcP-2fBHmvXe133dlW_fmKZes (Message 0 - eventsAfter)
	$cnVNVEDS7cSSN-m0rgfNAZOhfs8yey72rKbJf1zfe5U (Backfilled 2)
	$v2Iq9Ih4OyKReqJ2pcTDRb30HrrHEmG0qnhta47Crus (Backfilled 1)
	$njJCUsvrVaPtpyNP6PxW-PFuj9YQdowH3PVgC73wi44 (Backfilled 0)
	$hSNUYZq2RQVdjgRpXpq4EF4KnI1mU1eDdq5KncNkHak (Message 0 - eventsBefore)
]
```

Reply:
I guess each historical message could reference the `eventBefore` 🤔. It creates a forward extremities for each message imported. Seems dirty to break the nice clean chain of history. Are there any performance problems with this many extremities?

And the ordering just works because `stream_ordering` is decremented for each one.


Here is a diagram of how the DAG would look. Where `A <- B` are live events and historical `3` is imported first (reverse-chronological)

![](https://user-images.githubusercontent.com/558581/110751216-2c087380-8209-11eb-95bc-c6a3e09be240.png)

Reply:
> Actually diving into the Complement tests now, the reverse-chronological constraint seems like a problem. From the most recent historical message, I can't reference the previous historical message in `prev_event` because it hasn't been created yet

Oh, hum, that is annoying. I wonder if we just want an API that allows you to bulk insert events? I.e. have an API that takes multiple events and synapse takes care to Do The Right Thing (which would just be persisting them in one go rather than one by one)?

Reply:
For batching historical messages, we are still creating a forward extremity for every batch. **Is there any performance problem here?** Is making the DAG look sane a priority? (cc @erikjohnston) These seem like the main factors into whether we can use this `backfilled` approach. We would be doing the same DAG structure as single messages just with each one representing a whole batch and would run into the same problem just when we hit a magnitude more of messages.

> ![](https://user-images.githubusercontent.com/558581/110751216-2c087380-8209-11eb-95bc-c6a3e09be240.png)

---

Seems like the batching API would be it's own MSC. A batching version is mentioned in this draft API doc but it just says "A batching version of this API needs to be provided." for sending events, https://github.com/matrix-org/matrix-doc/blob/master/drafts/general_api.rst#send-a-message-draft

I was also looking for something to compare it to and was confused by seeing https://github.com/matrix-org/matrix-doc/pull/2244/files merged to add mass redactions and a `redacts` array to the `content`, but then it's not mentioned in the actual [spec](https://matrix.org/docs/spec/client_server/r0.6.1#redactions) at all.

I didn't see an MSC for batch sending events. Link welcome if it's actually a thing!


---

Stepping back from `backfilled` events, everything behaved well when we were using the explicit filter on `/sync` and importing chronological which also kept the DAG in a nice straight line.

Are there more reasons not to do it that way? `backfilled`, the name itself sounds correct to me for what we want to do but does complicate other areas.


---



#### Pro/con list


Chronological constraint:

 - Constraint of needing to import oldest messages -> most recent messages
 - Needs to be a normal message so that the `stream_ordering` is incremented (all `depth` is the same for historical messages so we tie-break based off of `stream_ordering`)
 - Can inherit `depth` straight from the `prev_events` defined which the code is already doing
 - Can point the `prev_event` of your next batch to your last one to make one long nice piece of history


Reverse chronological constraint:

 - Constraint of needing to import most recent messages -> oldest messages
 - Needs to use the `backfilled` option so that the `stream_ordering` is decremented  (all `depth` is the same for historical messages so we tie-break based off of `stream_ordering`)
 - Have to do a look-up for the successor events to so we can inherit their `depth` to get proper ordering and appear after the defined `prev_event`
 - From the most recent historical message, unable to reference the previous historical message in `prev_event` because it hasn't been created yet. The DAG is just a bunch of hanging extremity events hanging off the root event.
 - Mimics scrollback

Reply:
> Are there more reasons not to do it that way? `backfilled`, the name and definition of the word itself sounds correct to me for what we want to do but does complicate other areas.

Yeah, its mainly that I *know* that the `backfilled` flag is handled correctly everywhere, whereas we'd have to replicate that handling everywhere if we introduced a new mechanism.

> I was also looking for something to compare it to and was confused by seeing https://github.com/matrix-org/matrix-doc/pull/2244/files merged to add mass redactions and a `redacts` array to the `content`, but then it's not mentioned in the actual [spec](https://matrix.org/docs/spec/client_server/r0.6.1#redactions) at all.

It can take a while for MSCs to get written up into the spec itself, alas.



> For batching historical messages, we are still creating a forward extremity for every batch. **Is there any performance problem here?**

Thinking about it, we really don't want to reference this batch from any new events, as otherwise it'll cause other servers to pull the chunk in and treat it as "new" events, rather than old ones. We'll need to make sure when we're persisting that we don't add these events to the forward extremities table, which should be easily done (and that might already happen if they're marked as `backfilled`). To make this all work I think we really do want to ensure these batches of events are disconnected from the "main" DAG. 



> Seems like the batching API would be it's own MSC

I think its a reasonable thing to add to this MSC if we need it TBH, especially if it only gets used by ASes importing history. The API should be simple enough.

Reply:
> Thinking about it, we really don't want to reference this batch from any new events, as otherwise it'll cause other servers to pull the chunk in and treat it as "new" events, rather than old ones. 

That sounds like we aren't going to heal the state at all?

Also added my initial thoughts on historical state and membership events needed in a separate thread, https://github.com/matrix-org/synapse/pull/9247/files#r593351834


> We'll need to make sure when we're persisting that we don't add these events to the forward extremities table

Maybe I am misunderstanding the terminology. What would you call `1`, `2`, and `3` here? I assume `B` is a forward extremity and my understanding was each of the `1` -> `3` was as well.

> ![](https://user-images.githubusercontent.com/558581/110751216-2c087380-8209-11eb-95bc-c6a3e09be240.png)


> [batching] I think its a reasonable thing to add to this MSC if we need it

It seems like we can do without it and is just a nice extra to make things faster.


---

I am going to update the Complement tests to try out `backfill` and connecting each event to `eventBefore` to see how it works 🤞

Reply:
> > Thinking about it, we really don't want to reference this batch from any new events, as otherwise it'll cause other servers to pull the chunk in and treat it as "new" events, rather than old ones.
> 
> That sounds like we aren't going to heal the state at all?

Yeah, I think so. Probably, maybe. I think some of this is just going to have to shake out in practice.



> Maybe I am misunderstanding the terminology. What would you call `1`, `2`, and `3` in the image below? I assume `B` is a forward extremity and my understanding was each of the `1` -> `3` was as well.

So the "forwards extremities table" tracks what we view as the most recent extremities in the room, and which we use as prev events if we generate a new event. In your example, 1,2,3,B would all normally be added to the forward extremities table, and thus would be used as prev events for any new event. We also use the table to work out what the "current state" of the room is, by doing state res across all the states at the extremities. I don't think it really makes sense to add "historic" data to that.